File: sdk/python/gestalt/_resource.py
import dataclasses
from enum import Enum
from typing import Any


class DatastoreCapability(str, Enum):
    KEY_VALUE = "key_value"
    SQL = "sql"
    BLOB_STORE = "blob_store"


# ---------------------------------------------------------------------------
# Data types
# ---------------------------------------------------------------------------


@dataclasses.dataclass(slots=True)
class KVEntry:
    key: str = ""
    value: bytes = b""


@dataclasses.dataclass(slots=True)
class SQLRows:
    columns: list[str] = dataclasses.field(default_factory=list)
    rows: list[list[Any]] = dataclasses.field(default_factory=list)


@dataclasses.dataclass(slots=True)
class SQLExecResult:
    rows_affected: int = 0
    last_insert_id: int = 0


@dataclasses.dataclass(slots=True)
class SQLMigration:
    version: int
    description: str
    up_sql: str


@dataclasses.dataclass(slots=True)
class BlobEntry:
    key: str = ""
    size: int = 0
    content_type: str = ""
    last_modified: float | None = None


# ---------------------------------------------------------------------------
# Consumer interfaces (no namespace -- host injects it)
# ---------------------------------------------------------------------------


class KeyValueStore:
    def get(self, key: str) -> bytes | None:
        raise NotImplementedError

    def put(self, key: str, value: bytes) -> None:
        raise NotImplementedError

    def delete(self, key: str) -> None:
        raise NotImplementedError

    def list(self, _prefix: str = "") -> list[KVEntry]:
        raise NotImplementedError


class SQLStore:
    def query(self, _sql: str, _args: list[Any] | None = None) -> SQLRows:
        raise NotImplementedError

    def exec(self, _sql: str, _args: list[Any] | None = None) -> SQLExecResult:
        raise NotImplementedError

    def migrate(self, _migrations: list[SQLMigration]) -> None:
        raise NotImplementedError


class BlobStore:
    def get(self, _key: str) -> tuple[bytes, str] | None:
        raise NotImplementedError

    def put(self, _key: str, _data: bytes, _content_type: str = "") -> None:
        raise NotImplementedError

    def delete(self, _key: str) -> None:
        raise NotImplementedError

    def list(self, _prefix: str = "") -> list[BlobEntry]:
        raise NotImplementedError


# ---------------------------------------------------------------------------
# Provider interfaces (namespace-aware, implemented by resource providers)
# ---------------------------------------------------------------------------


class DatastoreProvider:
    def configure(self, _name: str, _config: dict[str, Any]) -> None:
        pass


class KeyValueDatastoreProvider(DatastoreProvider):
    def kv_get(self, _namespace: str, _key: str) -> bytes | None:
        raise NotImplementedError

    def kv_put(self, _namespace: str, _key: str, _value: bytes, _ttl_seconds: int = 0) -> None:
        raise NotImplementedError

    def kv_delete(self, _namespace: str, _key: str) -> None:
        raise NotImplementedError

    def kv_list(self, _namespace: str, _prefix: str = "", _cursor: str = "", _limit: int = 100) -> list[KVEntry]:
        raise NotImplementedError

    def kv_migrate(self, _namespace: str) -> None:
        raise NotImplementedError


class SQLDatastoreProvider(DatastoreProvider):
    def sql_query(self, _namespace: str, _sql: str, _args: list[Any] | None = None) -> SQLRows:
        raise NotImplementedError

    def sql_exec(self, _namespace: str, _sql: str, _args: list[Any] | None = None) -> SQLExecResult:
        raise NotImplementedError

    def sql_migrate(self, _namespace: str, _migrations: list[SQLMigration]) -> None:
        raise NotImplementedError


class BlobStoreDatastoreProvider(DatastoreProvider):
    def blob_get(self, _namespace: str, _key: str) -> tuple[bytes, str] | None:
        raise NotImplementedError

    def blob_put(self, _namespace: str, _key: str, _data: bytes, _content_type: str = "") -> None:
        raise NotImplementedError

    def blob_delete(self, _namespace: str, _key: str) -> None:
        raise NotImplementedError

    def blob_list(self, _namespace: str, _prefix: str = "") -> list[BlobEntry]:
        raise NotImplementedError

File: sdk/go/resource.go
package gestalt

import (
	"context"
	"time"
)

// DatastoreCapability identifies a storage capability that a resource provider
// can implement and that plugins can consume.
type DatastoreCapability string

const (
	CapabilityKeyValue  DatastoreCapability = "key_value"
	CapabilitySQL       DatastoreCapability = "sql"
	CapabilityBlobStore DatastoreCapability = "blob_store"
)

// KeyValueStore is the consumer interface for key-value storage. Plugins
// receive an implementation of this interface when they declare a key_value
// resource dependency. The namespace is injected by the host and is not
// exposed to consumers.
type KeyValueStore interface {
	Get(ctx context.Context, key string) (value []byte, found bool, err error)
	Put(ctx context.Context, key string, value []byte) error
	PutWithTTL(ctx context.Context, key string, value []byte, ttlSeconds int64) error
	Delete(ctx context.Context, key string) (deleted bool, err error)
	List(ctx context.Context, prefix string, cursor string, limit int32) (entries []KVEntry, nextCursor string, err error)
}

// KVEntry is a single key-value pair returned by [KeyValueStore.List].
type KVEntry struct {
	Key   string
	Value []byte
}

// SQLStore is the consumer interface for SQL storage. Plugins receive an
// implementation of this interface when they declare a sql resource dependency.
type SQLStore interface {
	Query(ctx context.Context, query string, params ...any) (*SQLRows, error)
	Exec(ctx context.Context, query string, params ...any) (SQLExecResult, error)
	Migrate(ctx context.Context, migrations []SQLMigration) error
}

// SQLRows holds the result of a SQL query.
type SQLRows struct {
	Columns []string
	Rows    [][]any
}

// SQLExecResult holds the result of a SQL exec statement.
type SQLExecResult struct {
	RowsAffected int64
	LastInsertID int64
}

// SQLMigration describes a single forward migration to be applied by the
// resource provider.
type SQLMigration struct {
	Version     int32
	Description string
	UpSQL       string
}

// BlobStore is the consumer interface for blob storage. Plugins receive an
// implementation of this interface when they declare a blob_store resource
// dependency.
type BlobStore interface {
	Get(ctx context.Context, key string) (data []byte, contentType string, err error)
	Put(ctx context.Context, key string, data []byte, contentType string) error
	Delete(ctx context.Context, key string) error
	List(ctx context.Context, prefix string, cursor string, limit int32) (entries []BlobEntry, nextCursor string, err error)
}

// BlobEntry describes a single blob returned by [BlobStore.List].
type BlobEntry struct {
	Key          string
	Size         int64
	ContentType  string
	LastModified time.Time
}
